File: src/main/java/com/google/cloud/dataproc/templates/gcs/GCStoBigquery.java
/*
 * Copyright (C) 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.google.cloud.dataproc.templates.gcs;

import com.google.cloud.dataproc.templates.BaseTemplate;
import com.google.cloud.dataproc.templates.options.BigQueryOutputOptions;
import com.google.cloud.dataproc.templates.options.GCSInputOptions;
import com.google.cloud.dataproc.templates.options.TemplateOptions;
import com.google.cloud.dataproc.templates.options.TemplateOptionsFactory;
import java.util.Objects;
import java.util.Properties;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SaveMode;
import org.apache.spark.sql.SparkSession;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class GCStoBigquery implements BaseTemplate {

  public static final Logger LOGGER = LoggerFactory.getLogger(GCStoBigquery.class);

  @Override
  public void runTemplate() {
    Properties props = getProperties();
    TemplateOptionsFactory<TemplateOptions> optionsFactory =
        TemplateOptionsFactory.fromProps(props);
    TemplateOptions options = optionsFactory.create();
    GCSInputOptions inputOptions = optionsFactory.as(GCSInputOptions.class).create();
    BigQueryOutputOptions outputOptions = optionsFactory.as(BigQueryOutputOptions.class).create();
    LOGGER.info("{}", options);
    LOGGER.info("{}", inputOptions);
    LOGGER.info("{}", outputOptions);

    SparkSession spark = null;
    try {
      spark = SparkSession.builder().appName("GCS to Bigquery load").getOrCreate();

      Dataset<Row> inputData = null;

      switch (inputOptions.getFormatEnum()) {
        case CSV:
          inputData =
              spark
                  .read()
                  .format("csv")
                  .option("header", true)
                  .option("inferSchema", true)
                  .load(inputOptions.getPath());
          break;
        case AVRO:
          inputData = spark.read().format("com.databricks.spark.avro").load(inputOptions.getPath());
          break;
        case PARQUET:
          inputData = spark.read().parquet(inputOptions.getPath());
          break;
        case JSON:
          inputData = spark.read().json(inputOptions.getPath());
          break;
        default:
          throw new IllegalArgumentException("Format not supported");
      }

      inputData
          .write()
          .format("com.google.cloud.spark.bigquery")
          .option("table", outputOptions.getTable())
          .option("temporaryGcsBucket", outputOptions.getTemporaryGcsBucket())
          .mode(SaveMode.Append)
          .save();

    } catch (Throwable th) {
      LOGGER.error("Exception in GCStoBigquery", th);
      if (Objects.nonNull(spark)) {
        spark.stop();
      }
    }
  }
}

File: python/src/templates/gcs/gcs_to_bigquery.py
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from pyspark.sql import SparkSession
import argparse, configparser

from template_constants import *

spark = SparkSession.builder.appName("GCS to Bigquery load").getOrCreate()
spark.sparkContext.setLogLevel("INFO")

log4jLogger = spark.sparkContext._jvm.org.apache.log4j
logger = log4jLogger.LogManager.getLogger(__name__)

def runTemplate():

  # Arguments
  args = parse_args()
  input_file_location = args[GCS_BQ_INPUT_LOCATION]
  big_query_dataset   = args[GCS_OUTPUT_DATASET_NAME]
  big_query_table     = args[GCS_OUTPUT_TABLE_NAME]
  input_file_format   = args[GCS_BQ_INPUT_FORMAT]
  bq_temp_bucket      = args[GCS_BQ_LD_TEMP_BUCKET_NAME]

  # Read
  if (input_file_format == GCS_BQ_PRQT_FORMAT):
    input_data = spark.read \
                      .parquet(input_file_location)
  elif (input_file_format == GCS_BQ_AVRO_FORMAT):
    input_data = spark.read \
                      .format(GCS_BQ_AVRO_EXTD_FORMAT) \
                      .load(input_file_location)
  elif (input_file_format == GCS_BQ_CSV_FORMAT):
    input_data = spark.read \
                      .format(GCS_BQ_CSV_FORMAT) \
                      .option(GCS_BQ_CSV_HEADER, True) \
                      .option(GCS_BQ_CSV_INFER_SCHEMA, True) \
                      .load(input_file_location)
  else:
    raise Exception("Currently avro, parquet and csv are the only supported formats")

  logger.info("Starting GCS to Bigquery spark job with following parameters\n 1. {}={}\n 2. {}={}\n 3. {}={}\n 4. {}={}\n 5. {}={}\n".format(
              GCS_BQ_INPUT_LOCATION,
              input_file_location,
              GCS_OUTPUT_DATASET_NAME,
              big_query_dataset,
              GCS_OUTPUT_TABLE_NAME,
              big_query_table,
              GCS_BQ_INPUT_FORMAT,
              input_file_format,
              GCS_BQ_LD_TEMP_BUCKET_NAME,
              bq_temp_bucket)
  )

  # Write
  input_data.write \
            .format(GCS_BQ_OUTPUT_FORMAT) \
            .option(GCS_BQ_OUTPUT, big_query_dataset + "." + big_query_table) \
            .option(GCS_BQ_TEMP_BUCKET, bq_temp_bucket) \
            .mode("append") \
            .save()

def parse_args():

  config = configparser.ConfigParser()
  config.read('default_args.ini')
  defaults = config['gcs_to_bigquery']

  parser = argparse.ArgumentParser()
  parser.add_argument('--'+GCS_BQ_INPUT_LOCATION, dest=GCS_BQ_INPUT_LOCATION)
  parser.add_argument('--'+GCS_OUTPUT_DATASET_NAME, dest=GCS_OUTPUT_DATASET_NAME)
  parser.add_argument('--'+GCS_OUTPUT_TABLE_NAME, dest=GCS_OUTPUT_TABLE_NAME)
  parser.add_argument('--'+GCS_BQ_INPUT_FORMAT, dest=GCS_BQ_INPUT_FORMAT)
  parser.add_argument('--'+GCS_BQ_LD_TEMP_BUCKET_NAME, dest=GCS_BQ_LD_TEMP_BUCKET_NAME)

  args = vars(parser.parse_args())
  result = dict(defaults)
  result.update({k: v for k, v in args.items() if v is not None})

  if (None in list(result.values())) or ("" in list(result.values())):
    raise Exception("Required parameters for gcs_to_bigquery not passed")

  return result

if __name__ == '__main__':
  runTemplate()